### ISSUE TITLE
Can't run correctly in my version

### ISSUE THREAD

**Issue by Jotakak-yu** [Ant-Brain/EfficientWord-Net-Trainer#1]
Hello, I have trained my own version model from the simple_trainer.py, but I found it not running correctly while the calculated confidence_score is  like this :
{'match': True, 'confidence': 1.779856}

and when I put the model file to Netron to check the model structure, I found some difference between your version: 

![屏幕截图 2024-02-28 215634](https://github.com/Ant-Brain/EfficientWord-Net-Trainer/assets/56828082/a40ea695-eb31-49f4-8dd9-16c7a57cfe0e)

 and mine:

![屏幕截图 2024-02-28 215651](https://github.com/Ant-Brain/EfficientWord-Net-Trainer/assets/56828082/783ef125-7c31-4212-a2c1-014d0efd5eea)

So I want to know is this difference causing the anomaly in confidence_scores? 
Can you please share your onnx conversion code? 
Or should I modify the scoreVector function in the audio_processing.py? How to modify?

**Comment by Jotakak-yu** [comment]
@TheSeriousProgrammer @aman-17

**Comment by TheSeriousProgrammer** [comment]
Hi did you pass the model outputs through softmax before the computation?

**Comment by Jotakak-yu** [comment]
Wow, according to your suggestion, I've added a softmax layer before the output, and it works! 

But I'm still curious why adding softmax makes it work correctly? Because this is not a classification problem in my opinion

Reason 1: According to your EfficientWord project, the working principle is to obtain a fixed-length embedding through the basemodel during inference to calculate the similarity.

Reason 2: It seems to me that your model output does not go through softmax, but it works fine

I'm new to deep learning, so it may seem like a stupid question to you :D
Anyway, it works correctly now, thank you very much for your help, thank you for such a good project！@TheSeriousProgrammer

**Comment by TheSeriousProgrammer** [comment]
My bad instead of softmax, add l2 normalisation as a post processing to your output and it should work fine

This works because both l2 normalisation and softmax pins the vector length to max of 1

These activation functions essentially boil down the vectors into the surface of a hyperdimensional sphere.

so when you perform cosine distance on the outputted vectors it will act as an unbounded classifier

Consider looking at the CLIP paper and arcface paper, it will help you understanding the basic concept of the latest model here.

Thanks for your kind words

**Comment by Jotakak-yu** [comment]
Thank you for your patient answer!

`feat_l2 = F.normalize(feat_norm, dim=1) `
Is this represent  l2 normalisation？If it is, I think I've already added it before, But it would produce such results without any utterance:
{'match': True, 'confidence': 0.987432}

**Comment by Jotakak-yu** [comment]
> My bad instead of softmax, add l2 normalisation as a post processing to your output and it should work fine
> 
> This works because both l2 normalisation and softmax pins the vector length to max of 1
> 
> These activation functions essentially boil down the vectors into the surface of a hyperdimensional sphere.
> 
> so when you perform cosine distance on the outputted vectors it will act as an unbounded classifier
> 
> Consider looking at the CLIP paper and arcface paper, it will help you understanding the basic concept of the latest model here.
> 
> Thanks for your kind words

Thank you for your patient answer!

`feat_l2 = F.normalize(feat_norm, dim=1) `
Is this represent l2 normalisation？If it is, I think I've already added it before, But it would produce such results without any utterance:
{'match': True, 'confidence': 0.987432}

**Comment by Jotakak-yu** [comment]
> > My bad instead of softmax, add l2 normalisation as a post processing to your output and it should work fine
> > This works because both l2 normalisation and softmax pins the vector length to max of 1
> > These activation functions essentially boil down the vectors into the surface of a hyperdimensional sphere.
> > so when you perform cosine distance on the outputted vectors it will act as an unbounded classifier
> > Consider looking at the CLIP paper and arcface paper, it will help you understanding the basic concept of the latest model here.
> > Thanks for your kind words
> 
> Thank you for your patient answer!
> 
> `feat_l2 = F.normalize(feat_norm, dim=1) ` Is this represent l2 normalisation？If it is, I think I've already added it before, But it would produce such results without any utterance: {'match': True, 'confidence': 0.987432}

Sorry I've made a mistake, I didn't switch to the corresponding model file during testing
So adding softmax doesn't work at all.

```
import torch
import torch.nn as nn
import torch.nn.functional as F
from simple_trainer import AudioClassifierVectorDatasetPL, LightningWordClassifier

class PartialModel(nn.Module):
    def __init__(self, original_model):
        super(PartialModel, self).__init__()
        self.inp_batch_norm = original_model.inp_batch_norm
        self.feature_network = original_model.feature_network
        self.bn_final = original_model.bn_final

    def forward(self, x):
        x_norm = self.inp_batch_norm(x)
        feat = self.feature_network(x_norm)
        feat_norm = self.bn_final(feat)
        feat_l2 = F.normalize(feat_norm, dim=1)
        # feat_output = F.softmax(feat_l2, dim=1)
        
        return feat_l2
    
pl_model = LightningWordClassifier()
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
pl_model = pl_model.to(device)
x = torch.randn(1, 1, 149, 64).to(device)

pl_model = LightningWordClassifier.load_from_checkpoint("/home/yu/project/Siamese-Resnet-Trainer/results_dir/res50/epoch=70-val_top1=0.8453.ckpt", map_location=device)  # Use the actual path of the ckpt
partial_model = PartialModel(pl_model.pytorch_model)  # Create partial model
torch.onnx.export(partial_model,
                  x,
                  'results_dir/test.onnx',
                  opset_version=13,
                  do_constant_folding=True,  # whether to execute constant folding optimization
                  input_names=["input"],  # input name
                  output_names=["output"],  # output name
                  )  # Export to ONNX
```
When I added l2 norm and converted the model to onnx, the problem still existed. @TheSeriousProgrammer

**Comment by TheSeriousProgrammer** [comment]
There are 2 possible sources of error over here

Dataset Issue
or
The final l2 regularization call would've not been exported as a part of the model . To verify the same check the length of the output vector by the length of the vector
or
Model issue wherein you are trying to train the model from scratch with few audio samples, this will result in overfitting

**Comment by Jotakak-yu** [comment]
> There are 2 possible sources of error over here
> 
> Dataset Issue or The final l2 regularization call would've not been exported as a part of the model . To verify the same check the length of the output vector by the length of the vector or Model issue wherein you are trying to train the model from scratch with few audio samples, this will result in overfitting

Firstly, regarding the issue with the dataset, I used the same complete dataset from MLCommons in English as you, so there is no problem of overfitting due to a small amount of data; 
Secondly, I used two types of features, logfbank and fbank, for training. In the final generated tar file, after my testing, the two types of audio information and spectrograms are as follows:

logfbank：

```
Key: anything_5663_0
URL: /home/panjianan/project/Siamese-Resnet-Trainer/dataset_conversion_scripts/shards_logfbank/test_vectors-000001.tar.gz
Wave data: [[-12.083953   -7.8507676  -8.764886  ... -14.614473  -14.498586
  -15.409697 ]
 [-10.361143   -9.018323  -10.401963  ... -14.52404   -14.678494
  -15.387562 ]
 [-13.5743     -9.031195  -11.267475  ... -14.748161  -14.258674
  -15.096966 ]
 ...
 [-10.873442   -8.822695   -9.33332   ... -16.148117  -15.951692
  -16.60814  ]
 [ -9.723149  -11.649402   -9.016218  ... -15.511104  -15.504323
  -16.137627 ]
 [-11.220969  -11.688125   -8.745978  ... -15.939489  -16.025082
  -16.594007 ]]
Label: tensor([99.])
```

![Figure_1](https://github.com/Ant-Brain/EfficientWord-Net-Trainer/assets/56828082/14d4326c-1630-4b7f-918f-7789901a25a4)

fbank：

```
Key: before_8755_0
URL: /home/panjianan/project/Siamese-Resnet-Trainer/dataset_conversion_scripts/shards_smallt/test_vectors-000001.tar.gz
Wave data: [[1.06066909e-11 1.42719321e-08 1.32981981e-08 ... 7.24549309e-08
  9.03408832e-08 1.85726226e-08]
 [3.96270172e-09 6.03435080e-09 7.87630672e-09 ... 5.31224060e-08
  8.77136017e-08 5.63216274e-09]
 [1.99749750e-09 1.54387314e-09 3.91290289e-09 ... 7.07343020e-08
  7.16264879e-08 1.41900474e-08]
 ...
 [2.82400303e-09 5.70053205e-09 2.14415019e-09 ... 1.53360524e-08
  2.07183959e-08 7.25613525e-09]
 [4.83573286e-11 1.06732900e-09 1.10729985e-08 ... 2.38039757e-08
  1.75024368e-08 7.64963204e-09]
 [6.22408071e-12 2.50540877e-09 1.88194509e-08 ... 1.49366741e-08
  2.02565182e-08 9.88658577e-09]]
Label: tensor([185.])
```
![Figure_2](https://github.com/Ant-Brain/EfficientWord-Net-Trainer/assets/56828082/e79e3d9b-3af4-4af1-8c04-5ab901bf343c)

Then for l2 regularization: When I export the onnx model, adding or not adding l2 regularization has an impact on the calculation of confidence. If not added, a confidence level greater than 1 will be obtained
During model inference, I printed the information for inp_vector and embedding as follows:
Using your model:
```
(1, 2048) (7, 2048)
[[ 0.01204111  0.00400959  0.03304974 ... -0.03338393 -0.00022337
  -0.00636036]] [[ 3.68688926e-02 -9.56422370e-03  3.54975760e-02 ...  4.19836817e-03
  -1.66153274e-02  9.63276034e-05]
 [ 1.53056402e-02 -1.27893360e-02  1.68258660e-02 ...  1.03832055e-02
  -8.98275524e-03 -1.57411583e-02]
 [ 2.63231993e-02 -1.74490511e-02  1.97591260e-02 ... -7.21763645e-04
  -3.69803282e-03 -7.45239016e-03]
 ...
 [ 2.06869412e-02  9.21848696e-03  7.23458128e-03 ...  2.10485980e-02
  -1.94992535e-02  3.01778857e-02]
 [ 3.58835571e-02 -3.31663899e-03  1.92190949e-02 ...  6.85139978e-03
  -8.50650854e-03  2.61208192e-02]
 [ 2.08313912e-02 -3.76849179e-03  1.67783368e-02 ...  1.97250079e-02
  -1.51222693e-02 -1.80331767e-02]]
[[-0.01681481]
 [-0.05153824]
 [-0.08911458]
 [-0.06376237]
 [-0.02864286]
 [-0.05624036]
 [-0.06013595]]
0.4915926
(7, 1) -0.01681481
```
and mine：
```
(1, 1152) (7, 1152)
[[-0.00924187 -0.02104287 -0.02507884 ... -0.01886808  0.00479398
  -0.00147296]] [[-1.13070374e-02 -1.71661340e-02 -2.17153337e-02 ... -2.03265026e-02
   6.68438384e-03  3.22929211e-03]
 [-1.54486606e-02 -1.63398143e-02 -2.34954450e-02 ... -2.03943346e-02
  -8.90998708e-05  2.17664102e-03]
 [-1.02941701e-02 -1.51285613e-02 -2.19966527e-02 ... -2.03330629e-02
   7.49744382e-03 -1.15562021e-03]
 ...
 [-1.55086825e-02 -1.55044198e-02 -2.26066951e-02 ... -2.04412024e-02
   2.86136474e-03  1.33222889e-03]
 [-9.14190058e-03 -1.88715067e-02 -2.08062269e-02 ... -2.05804035e-02
   6.18852349e-03  2.81215366e-03]
 [-1.46094058e-02 -1.78781953e-02 -2.23490670e-02 ... -2.02713348e-02
   8.33997224e-03  1.28860900e-03]]
[[0.9851996 ]
 [0.9757447 ]
 [0.9823964 ]
 [0.9776782 ]
 [0.97612166]
 [0.98778856]
 [0.98651457]]
0.9938943
(7, 1) 0.98778856
{'match': True, 'confidence': 0.9938943}
```
I don't know why my model produces such a high cosine similarity without any utterance

**Comment by TheSeriousProgrammer** [comment]
I would suggest you to debug the data pipeline, check the output datapoints from the dataloader for the train and test pipelines
